### PR TITLE
ollama: Support model context_size (num_ctx) >2048

### LINKED ISSUE #15200: Requesting ability to set max_tokens for ollama models
### Check for existing issues

- [X] Completed

### Describe the feature

There doesn't seem to be any way to set the `max_tokens` for an ollama assistant model with the 0.147.0 nightly I'm using. I've attempted to set it in the config in a variety of ways and none of them had any visible impact in the UI. I always see "_n_ / 2k" in the top right corner, no matter which model I select.

### If applicable, add mockups / screenshots to help present your vision of the feature

image

### PR DESCRIPTION
**Pull request by notpeter**
Closes #15200

Hard code context windows for the most popular models.  In the future we should probably allow overriding this via settings or determining it dynamically via API.

Depending on hardware, users may wish to increase their timeout when work :
```
  "language_models": {
    "ollama": {
      "low_speed_timeout_in_seconds": 120
    }
```

Release Notes:

- Added support for Ollama models with context length >2048.

Screenshot 2024-08-26 at 10 18 46

### CODE REVIEW

File: assets/settings/default.json [hunk 1]
```diff
@@ -895,7 +895,8 @@
       "api_url": "https://generativelanguage.googleapis.com"
     },
     "ollama": {
-      "api_url": "http://localhost:11434"
+      "api_url": "http://localhost:11434",
+      "low_speed_timeout_in_seconds": 60
```
Comment (rgbkrk) [inline assets/settings/default.json:899]: 💯 

People almost _always_ need this.

File: crates/ollama/src/ollama.rs [hunk 2]
```diff
@@ -70,11 +70,36 @@ pub struct Model {
...
+            | "qwen" | "qwen2" | "wizardlm2" | "wizard-math" => 32768,
+            "codellama" | "stable-code" | "deepseek-coder" | "starcoder2"   // 16K
+            | "wizardcoder" => 16384,
+            "llama3" | "gemma2" | "gemma" | "codegemma" | "dolphin-llama3"  // 8K
+            | "llava-llama3" | "starcoder" | "openchat" | "aya" => 8192,
+            "llama2" | "yi" | "llama2-chinese" | "vicuna" | "nous-hermes2"  // 4K
+            | "stablelm2" => 4096,
+            "phi" | "orca-mini" | "tinyllama" | "granite-code" => 2048,     // 2K
+            _ => 2048,                                                      // 2K (default)
+        },
```
Comment (rgbkrk) [inline crates/ollama/src/ollama.rs:94]: Rad, I think this is great.

Comment (rgbkrk) [reply crates/ollama/src/ollama.rs:94]: It's going to be slightly wrong still as the tokenization is different between many of these models and tiktoken itself. However, it's a good estimate for the user.

**Review by rgbkrk — APPROVED** [review]

File: crates/ollama/src/ollama.rs (hunk 2 above)
Comment (notpeter) [reply crates/ollama/src/ollama.rs:94]: The estimates are still fuzzy, but previously our requests was including `num_ctx` 2048 which caused ollama to prematurely truncate context there.

Comment (rgbkrk) [reply crates/ollama/src/ollama.rs:94]: Oh yikes, I didn't realize this was truncating it! I thought it was just our estimate for the window. Thank you!

Comment (notpeter) [reply crates/ollama/src/ollama.rs:94]: Yep.
```sh
# tail -f ~/.ollama/logs/server.log
INFO [update_slots] input truncated | n_ctx=2048 n_erase=74657 n_keep=4 n_left=2044 n_shift=1022 tid="0x1f2dccf40" timestamp=1724689205
```

**Comment by rgbkrk** [comment]
:shipit:

**Pull request merged**